WooCommerce/Classes/ViewRelated/Orders/Collect Payments/CollectOrderPaymentUseCase.swift

```diff
@@ -451,6 +429,51 @@ private extension CollectOrderPaymentUseCase {
         )
     }
 
+    private func presentRetryWithoutRestartingError(error: Error,
+                                                    paymentAlerts: CardReaderTransactionAlertsProviding,
+                                                    paymentGatewayAccount: PaymentGatewayAccount,
+                                                    onCompletion: @escaping (Result<CardPresentCapturedPaymentData, Error>) -> ()) {
+        alertsPresenter.present(
+            viewModel: paymentAlerts.error(
+                error: error,
+                tryAgain: { [weak self] in
+                    guard let self = self else { return }
+                    self.checkOrderIsStillEligibleForPayment(alertProvider: paymentAlerts, onPaymentCompletion: onCompletion ) { result in
```

Comment:
I'm not sure about this: Do we need to "weakify self" further down the function here? Since we have nested closures it would seem that we can create reference cycles on lines 445 and 452, however, if we do, I understand that we may want to keep the strong reference for the retry logic?

Reply:
Hmm. I don't _think_ this will cause a retain cycle, but I'm struggling to 100% convince myself.

We don't store of the nested closures in a way which would cause a retain cycle, but of course that could change over time, and where they're passed to to publishers in the CardReaderService my reasoning becomes less sure.

I think it's fine in the places you mention; the alerts are what's retaining `onCompletion`, and they only last as long as they're displayed. 

The `CardPresentPaymentAction.retryPayment` also holds it, and the `StripeCardReaderService` in turn holds it in a publisher chain, which maybe could cause a retain cycle as that service vends the publisher chains back out to Yosemite, which then get observed by classes like the PaymentOrchestrator – I've not been able to cause that, but I've put in a weak/strong dance to match the `collectPayment` implementation just in case. See d99a9ca